File: src/main/java/Timeline.java
import org.json.JSONArray;
import org.json.JSONObject;
import spark.Request;
import spark.Response;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;
import java.text.ParseException;

public class Timeline {
    ConfigItem configItem;
    public Timeline(ConfigItem configItem) {
        this.configItem = configItem;
        //getData(1597968000, 1598054400);
    }

    public String getDataDate(Request request, Response response) {
        String res = "";
        java.util.Date dt;

        java.text.SimpleDateFormat sdf =
                new java.text.SimpleDateFormat("yyyy-MM-dd");

        try {
            dt = sdf.parse(request.queryParams("date"));
            res = getData(dt.getTime() / 1000, (dt.getTime() + 86400000) / 1000).toString();
        } catch (ParseException e) {
            System.out.println("asodf");
            e.printStackTrace();
        }
        System.out.println(res);
        return res;
    }

    private JSONObject getData(long start, long end) {
        JSONArray jsonArray = new JSONArray();
        JSONArray jsonArrayAll = new JSONArray();
        JSONArray jsonArrayRoutes = new JSONArray();
        JSONObject jsonObjectRes = new JSONObject();

        try {
            Statement stmt = Mysql.conn.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM data WHERE date BETWEEN '" + Mqtt.getMysqlDateString(start) + "' AND '"
                    + Mqtt.getMysqlDateString(end) + "'");
            double lat = 0;
            double lon = 0;
            long time = 0;
            boolean multiple = true;
            boolean added = true;
            int count = 0;
            double latTot = 0;
            double lonTot = 0;
            Timestamp firstTime = null;
            Timestamp endTime = null;
            while (rs.next()) {
                if (distance(lat, rs.getDouble("lat"), lon, rs.getDouble("lon"), 0, 0) >= configItem.getRadiusLocation()) {
                    time = rs.getTimestamp("date").getTime();
                    multiple = false;
                    if (!added) {
                        jsonArray.put(add(latTot, lonTot, count, firstTime, endTime));
                        added = true;
                    }
                    count = 0;
                    latTot = 0;
                    lonTot = 0;
                    firstTime = rs.getTimestamp("date");
                    lat = rs.getDouble("lat");
                    lon = rs.getDouble("lon");
                }
                if (distance(lat, rs.getDouble("lat"), lon, rs.getDouble("lon"), 0, 0) < configItem.getRadiusLocation()) {
                    if (rs.getTimestamp("date").getTime() - time > 420000 && !multiple) {

                        multiple = true;
                        added = false;
                    }
                    count += 1;
                    latTot = latTot + rs.getDouble("lat");
                    lonTot = lonTot + rs.getDouble("lon");
                    endTime = rs.getTimestamp("date");
                }
                JSONObject jsonObject = new JSONObject();
                jsonObject.put("date", rs.getTimestamp("date").getTime());
                jsonObject.put("lat", rs.getDouble("lat"));
                jsonObject.put("lon", rs.getDouble("lon"));
                jsonArrayAll.put(jsonObject);
            }
            if (!added) {
                jsonArray.put(add(latTot, lonTot, count, firstTime, endTime));
            }
            rs.close();
            stmt.close();
            jsonArrayRoutes = Routes.getRouteFromStop(jsonArray, jsonArrayAll);
        } catch (SQLException e) {
            e.printStackTrace();
        }
        jsonObjectRes.put("routes", jsonArrayRoutes);
        jsonObjectRes.put("stops", jsonArray);

        return jsonObjectRes;
    }

    public JSONObject add(double latTot, double lonTot, int count, Timestamp firstTime, Timestamp endTime) {
        JSONObject jsonObjectLoc = new JSONObject();
        Stop stop = StopHandler.getAddressName(round(latTot / count, 5), round(lonTot / count, 5), configItem);
        jsonObjectLoc.put("start", firstTime.getTime());
        jsonObjectLoc.put("end", endTime.getTime());
        jsonObjectLoc.put("location", stop.getName());
        jsonObjectLoc.put("locationUserAdded", stop.isUserAdded());
        jsonObjectLoc.put("lat", round(latTot / count, 5));
        jsonObjectLoc.put("lon", round(lonTot / count, 5));
        return jsonObjectLoc;
    }

    public static double distance(double lat1, double lat2, double lon1,
                                  double lon2, double el1, double el2) {

        final int R = 6371; // Radius of the earth

        double latDistance = Math.toRadians(lat2 - lat1);
        double lonDistance = Math.toRadians(lon2 - lon1);
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        double distance = R * c * 1000; // convert to meters

        double height = el1 - el2;

        distance = Math.pow(distance, 2) + Math.pow(height, 2);

        return Math.sqrt(distance);
    }

    public static double round(double value, int places) {
        if (places < 0) throw new IllegalArgumentException();

        BigDecimal bd = new BigDecimal(value);
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }
}

File: src/main/java/Timeline.kt
import data.LocationDataDaoImpl
import model.Location
import model.LocationItem
import org.json.JSONArray
import org.json.JSONObject
import java.lang.Exception
import java.math.BigDecimal
import java.math.RoundingMode
import java.sql.Timestamp
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.*

class Timeline(val configItem: ConfigItem, val mysql: Mysql) {
    fun getDataDate(dt: Date): String {
        return getData(mysql.locationDataDao.getData(dt.time / 1000, (dt.time + 86400000) / 1000)).toString()
    }

    fun getData(locationItems: ArrayList<LocationItem>): JSONObject {
        val jsonArray = JSONArray()
        val jsonArrayAll = JSONArray()
        val jsonObjectRes = JSONObject()

        var lat = 0.0
        var lon = 0.0
        var time: Long = 0
        var multiple = true
        var added = true
        var count = 0
        var latTot = 0.0
        var lonTot = 0.0
        var firstTime: Timestamp? = null
        var endTime: Timestamp? = null

        locationItems.forEach { item ->
            if (distance(lat, item.lat, lon, item.lon, 0.0, 0.0) >= configItem.radiusLocation) {
                time = item.date.time
                multiple = false
                if (!added) {
                    jsonArray.put(add(latTot, lonTot, count, firstTime!!, endTime!!))
                    added = true
                }
                count = 0
                latTot = 0.0
                lonTot = 0.0
                firstTime = item.date
                lat = item.lat
                lon = item.lon
            }
            if (distance(lat, item.lat, lon, item.lon, 0.0, 0.0) < configItem.radiusLocation) {
                if (item.date.time - time > 420000 && !multiple) {
                    multiple = true
                    added = false
                }
                count += 1
                latTot += item.lat
                lonTot += item.lon
                endTime = item.date
            }
            jsonArrayAll.put(JSONObject().apply {
                put("date", item.date.time)
                put("lat", item.lat)
                put("lon", item.lon)
            })
        }
        if (!added) {
            jsonArray.put(add(latTot, lonTot, count, firstTime!!, endTime!!))
        }

        val jsonArrayRoutes = Routes().getRouteFromStop(jsonArray, jsonArrayAll)

        //remove parts with possible inaccurate gps data
        try {
            var index = 0
            var lastIndex = jsonArrayRoutes.length() - 1

            //loop through all routes
            while (index <= lastIndex && lastIndex >= 0) {
                val item = jsonArrayRoutes.getJSONObject(index)
                if (item.getInt("pointCount") <= 4 && item.getDouble("distance") < 400 &&
                        item.getString("startLocation") == item.getString("stopLocation")
                ) {
                    for (k in 0 until jsonArray.length() - 1) {
                        val itemStop = jsonArray.getJSONObject(k)
                        if (itemStop.getLong("end") == item.getLong("start")) {//get the stop before the route that is being removed
                            jsonArrayRoutes.remove(index)
                            jsonArray.getJSONObject(k).put("end", jsonArray.getJSONObject(k + 1).getLong("end"))
                            jsonArray.remove(k + 1)
                            lastIndex--
                        }
                    }
                }
                index++
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        jsonObjectRes.put("routes", jsonArrayRoutes)
        jsonObjectRes.put("stops", jsonArray)

        return jsonObjectRes
    }

    fun addItemsToDb() {
        val locationsDb = mysql.locationDao.getLocations(true)
        if (locationsDb.size == 0) {
            // add everything to db
            println("Adding all locations to db")
            val res = getData(mysql.locationDataDao.getData(0))
            val jsonArray = res.getJSONArray("stops")
            for (g in 0 until jsonArray.length()) {
                val itemL = jsonArray.getJSONObject(g)
                mysql.locationDao.addLocation(itemL)
            }
        } else {
            println("Adding new locations to db")
            val res = getData(mysql.locationDataDao.getData(locationsDb[0].startDate.time / 1000))
            val jsonArray = res.getJSONArray("stops")
            val item = jsonArray.getJSONObject(0)
            mysql.locationDao.updateLocation(Location(
                    locationsDb[0].locationId,
                    Timestamp(item.getLong("start")),
                    Timestamp(item.getLong("end")),
                    item.getInt("stopId")
            ))

            jsonArray.remove(0)
            for (g in 0 until jsonArray.length()) {
                val itemL = jsonArray.getJSONObject(g)
                mysql.locationDao.addLocation(itemL)
            }
        }
    }

    fun add(latTot: Double, lonTot: Double, count: Int, firstTime: Timestamp, endTime: Timestamp): JSONObject {
        val stop = Address().getAddressName(
                round(latTot / count, 5),
                round(lonTot / count, 5),
                configItem,
                mysql
        )

        return JSONObject().apply {
            put("start", firstTime.time)
            put("end", endTime.time)
            if (stop.customName == null) {
                put("location", stop.name)
            } else {
                put("location", stop.customName)
            }
            put("lat", round(latTot / count, 5))
            put("lon", round(lonTot / count, 5))
            put("stopId", stop.stopId)
        }
    }

    companion object {
        fun distance(lat1: Double, lat2: Double, lon1: Double, lon2: Double, el1: Double, el2: Double): Double {
            val r = 6371 // Radius of the earth
            val latDistance = Math.toRadians(lat2 - lat1)
            val lonDistance = Math.toRadians(lon2 - lon1)
            val a = (sin(latDistance / 2) * sin(latDistance / 2)
                    + (cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2))
                    * sin(lonDistance / 2) * sin(lonDistance / 2)))
            val c = 2 * atan2(sqrt(a), sqrt(1 - a))
            var distance = r * c * 1000 // convert to meters
            val height = el1 - el2
            distance = distance.pow(2.0) + height.pow(2.0)
            return sqrt(distance)
        }

        fun round(value: Double, places: Int): Double {
            require(places >= 0)
            var bd = BigDecimal(value)
            bd = bd.setScale(places, RoundingMode.HALF_UP)
            return bd.toDouble()
        }
    }
}